### ISSUE TITLE
One instance crashed in Rethinkdb 2.2.5 cluster, possible broken data

### ISSUE THREAD

**Issue by ntquyen** [rethinkdb/rethinkdb#6033]
I'm running rethinkdb 2.2.5 cluster of 3 instances:

- OS: coreos 899.15.0 (Linux coreos-data-4 4.3.6-coreos #2 SMP Tue Apr 5 10:32:16 UTC 2016 x86_64 Intel(R) Xeon(R) CPU E5-2630 0 @ 2.30GHz GenuineIntel GNU/Linux)
- Docker 1.9.1

One instance in the cluster suddenly crashed and cannot be restarted, backtrace: 

```
Running rethinkdb 2.2.5~0jessie (GCC 4.9.2)...
Running on Linux 4.3.6-coreos x86_64
Loading data from directory /data
Listening for intracluster connections on port 29015
Connected to server "4200b06bc82a_1yx" 83a15f4a-a772-4649-b866-d537e5c0cfba
Connected to proxy ee2e9ad4-3bd7-48a0-ba3f-2c9406ac3d02
Connected to server "02f81613b7c1_wdx" 8138c225-e59c-4e5c-8f2e-93a5689aea47
Listening for client driver connections on port 28015
Listening for administrative HTTP connections on port 8080
Listening on addresses: 127.0.0.1, 10.2.44.2, ::1, fe80::42:aff:fe02:2c02%5
Server ready, "c1d17389be71_29y" d5a8f7d7-33b2-4f60-9561-38d4a9d163e1
A newer version of the RethinkDB server is available: 2.3.4. You can read the changelog at <https://github.com/rethinkdb/rethinkdb/releases>.
Version: rethinkdb 2.2.5~0jessie (GCC 4.9.2)
error: Error in src/rdb_protocol/serialize_datum.cc at line 814:
error: Guarantee failed: [datum_deserialize(&read_stream, &type) == archive_result_t::SUCCESS] Deserialization of datum type from buf failed with error archive_result_t::RANGE_ERROR.
error: Backtrace:
error: Fri Aug  5 03:52:04 2016

       1 [0xa8b7f0]: backtrace_t::backtrace_t() at 0xa8b7f0 (rethinkdb)
       2 [0xa8bb69]: format_backtrace(bool) at 0xa8bb69 (rethinkdb)
       3 [0xd0135d]: report_fatal_error(char const*, int, char const*, ...) at 0xd0135d (rethinkdb)
       4 [0x92f080]: ql::datum_deserialize_from_buf(shared_buf_ref_t<char> const&, unsigned long) at 0x92f080 (rethinkdb)
       5 [0x92f656]: ql::datum_deserialize_pair_from_buf(shared_buf_ref_t<char> const&, unsigned long) at 0x92f656 (rethinkdb)
       6 [0x8d3329]: ql::datum_t::unchecked_get_pair(unsigned long) const at 0x8d3329 (rethinkdb)
       7 [0x8d60fb]: rethinkdb() [0x8d60fb] at 0x8d60fb ()
       8 [0x8d61d0]: ql::datum_t::is_ptype() const at 0x8d61d0 (rethinkdb)
       9 [0x87668b]: ql::obj_or_seq_op_impl_t::eval_impl_dereferenced(ql::term_t const*, ql::scope_env_t*, ql::args_t*, scoped_ptr_t<ql::val_t> const&, std::function<scoped_ptr_t<ql::val_t> ()>) const at 0x87668b (rethinkdb)
       10 [0x87952f]: ql::bracket_term_t::eval_impl(ql::scope_env_t*, ql::args_t*, ql::eval_flags_t) const at 0x87952f (rethinkdb)
       11 [0x7d0952]: ql::op_term_t::term_eval(ql::scope_env_t*, ql::eval_flags_t) const at 0x7d0952 (rethinkdb)
       12 [0x8ea64a]: ql::runtime_term_t::eval_on_current_stack(ql::scope_env_t*, ql::eval_flags_t) const at 0x8ea64a (rethinkdb)
       13 [0x8ea94c]: ql::runtime_term_t::eval(ql::scope_env_t*, ql::eval_flags_t) const at 0x8ea94c (rethinkdb)
       14 [0x7cd19a]: ql::op_term_t::maybe_grouped_data(ql::scope_env_t*, ql::argvec_t*, ql::eval_flags_t, counted_t<ql::grouped_data_t>*, scoped_ptr_t<ql::val_t>*) const at 0x7cd19a (rethinkdb)
       15 [0x7d030b]: ql::op_term_t::term_eval(ql::scope_env_t*, ql::eval_flags_t) const at 0x7d030b (rethinkdb)
       16 [0x8ea64a]: ql::runtime_term_t::eval_on_current_stack(ql::scope_env_t*, ql::eval_flags_t) const at 0x8ea64a (rethinkdb)
       17 [0x8ea94c]: ql::runtime_term_t::eval(ql::scope_env_t*, ql::eval_flags_t) const at 0x8ea94c (rethinkdb)
       18 [0x80ddcb]: ql::reql_func_t::call(ql::env_t*, std::vector<ql::datum_t, std::allocator<ql::datum_t> > const&, ql::eval_flags_t) const at 0x80ddcb (rethinkdb)
       19 [0x80d436]: ql::func_t::call(ql::env_t*, ql::datum_t, ql::eval_flags_t) const at 0x80d436 (rethinkdb)
       20 [0x7b8383]: compute_keys(store_key_t const&, ql::datum_t, sindex_disk_info_t const&, std::vector<std::pair<store_key_t, ql::datum_t>, std::allocator<std::pair<store_key_t, ql::datum_t> > >*) at 0x7b8383 (rethinkdb)
       21 [0x7b9fb3]: rdb_update_single_sindex(store_t*, store_t::sindex_access_t const*, deletion_context_t const*, rdb_modification_report_t const*, unsigned long*, auto_drainer_t::lock_t, cond_t*, std::vector<std::pair<ql::datum_t, boost::optional<unsigned long> >, std::allocator<std::pair<ql::datum_t, boost::optional<unsigned long> > > >*, std::vector<std::pair<ql::datum_t, boost::optional<unsigned long> >, std::allocator<std::pair<ql::datum_t, boost::optional<unsigned long> > > >*) at 0x7b9fb3 (rethinkdb)
       22 [0x7c2559]: callable_action_instance_t<std::_Bind<void (*(store_t*, store_t::sindex_access_t*, deletion_context_t const*, rdb_modification_report_t const*, unsigned long*, auto_drainer_t::lock_t, cond_t*, std::vector<std::pair<ql::datum_t, boost::optional<unsigned long> >, std::allocator<std::pair<ql::datum_t, boost::optional<unsigned long> > > >*, std::vector<std::pair<ql::datum_t, boost::optional<unsigned long> >, std::allocator<std::pair<ql::datum_t, boost::optional<unsigned long> > > >*))(store_t*, store_t::sindex_access_t const*, deletion_context_t const*, rdb_modification_report_t const*, unsigned long*, auto_drainer_t::lock_t, cond_t*, std::vector<std::pair<ql::datum_t, boost::optional<unsigned long> >, std::allocator<std::pair<ql::datum_t, boost::optional<unsigned long> > > >*, std::vector<std::pair<ql::datum_t, boost::optional<unsigned long> >, std::allocator<std::pair<ql::datum_t, boost::optional<unsigned long> > > >*)> >::run_action() at 0x7c2559 (rethinkdb)
       23 [0x99c958]: coro_t::run() at 0x99c958 (rethinkdb)
error: Exiting.
```
When the crash occurred, `systemd-coredum` took all the CPU and cannot be stopped unless I stop the rethinkdb container and restart the server. After the server restarted and `systemd-coredum` disabled, rethinkdb still cannot be started, same crash.

It is likely the data is somehow broken (Deserialization failed). I'm have 10 rethinkdb clusters running in production but this is the first time I got this issue.

**Comment by danielmewes** [comment]
That definitely looks like a data corruption issue. Do you have a backup of this data (or another replica)?

Have you recently created any new secondary indexes on this cluster?
In RethinkDB 2.3.2 we've fixed a data corruption issue related to index construction https://github.com/rethinkdb/rethinkdb/issues/5715
If you've constructed any indexes recently, this could be related.

In that case you could also be able to recover the system by dropping all secondary indexes and then recreating them while the system is not under load. This assumes of course that you can at least start the server and connect to it without it crashing when it's not under load.

Does it crash immediately after you restart it? The stacktrace suggests that the crash happens in response to either a write operation, or an index construction (which would be resumed automatically right after a restart). Backfilling could also enter this code-path when the server reconnects to other replicas.

**Comment by ntquyen** [comment]
@danielmewes Fortunately, I got replicas in other two instances, so probably no data loss. I replaced the crash instance with a new one and rebalanced.

"Have you recently created any new secondary indexes on this cluster?"
--> No, this cluster has been running for a long time, no changes has been made recently. One thing I concern that this cluster contains very big data size (220 GB, 3 shards, 3 replicas for every tables) while the cache size assigned for total cluster is quite small, 2GB/instance)

"Does it crash immediately after you restart it?"
--> Yes, a few seconds after "Server ready" status

**Comment by danielmewes** [comment]
@ntquyen Do you still have the broken data folder?
Would you be able to upload a copy of the data to us for further inspection?
You can try tar-gzipping it and see how big the resulting file is. Often the data files compress pretty well. We can give you access to an upload server for uploading the archive then.

**Comment by ntquyen** [comment]
That's one of our biggest cluster which is around 80GB per instance, sorry I had to delete it to free up disk space...

**Comment by danielmewes** [comment]
I might have managed to reproduce this, but it's still intermittent.

I set up a cluster of three servers. Then I created 200 tables, sharded and replicated them, and inserted a bunch of documents into each. I then shut down and restarted some of the servers a few times, sometimes killing them in the middle of shutdown (not sure if that's relevant). At some point, I inserted some more documents while one of the three servers was down (not sure if that's relevant).
I was running all servers with the `--direct-io` option and on rotational disks.

Eventually I got data corruption:
```
error: Error in src/rdb_protocol/lazy_btree_val.cc at line 22:
error: Guarantee failed: [res == archive_result_t::SUCCESS] Deserialization of rdb value failed with error archive_result_t::RANGE_ERROR.
error: Backtrace:
error: Fri Aug 12 13:47:18 2016

       1 [0xb36390]: backtrace_t::backtrace_t() at 0xb36390 (rethinkdb)
       2 [0xb36723]: format_backtrace(bool) at 0xb36723 (rethinkdb)
       3 [0xd9051d]: report_fatal_error(char const*, int, char const*, ...) at 0xd9051d (rethinkdb)
       4 [0x9d6183]: get_data(rdb_value_t const*, buf_parent_t) at 0x9d6183 (rethinkdb)
       5 [0x8cd279]: post_construct_traversal_helper_t::handle_pair(scoped_key_value_t&&, concurrent_traversal_fifo_enforcer_signal_t) at 0x8cd279 (rethinkdb)
       6 [0xab98f1]: concurrent_traversal_adapter_t::handle_pair_coro(scoped_key_value_t*, semaphore_acq_t*, fifo_enforcer_write_token_t, auto_drainer_t::lock_t) at 0xab98f1 (rethinkdb)
       7 [0xab9674]: callable_action_instance_t<std::_Bind<std::_Mem_fn<void (concurrent_traversal_adapter_t::*)(scoped_key_value_t*, semaphore_acq_t*, fifo_enforcer_write_token_t, auto_drainer_t::lock_t)> (concurrent_traversal_adapter_t*, scoped_key_value_t*, semaphore_acq_t*, fifo_enforcer_write_token_t, auto_drainer_t::lock_t)> >::run_action() at 0xab9674 (rethinkdb)
       8 [0xa56fc7]: coro_t::run() at 0xa56fc7 (rethinkdb)
error: Exiting.
```

**Comment by danielmewes** [comment]
I also created a bunch of secondary indexes in my latest run, but those don't seem to be necessary for reproducing the issue (I got it before without indexes). With the indexes though, the crash happens quickly after restarting the server, just as @ntquyen reported:
```
Version: rethinkdb 2.3.4~0precise (GCC 4.8.1)
error: Error in src/rdb_protocol/lazy_btree_val.cc at line 22:
error: Guarantee failed: [res == archive_result_t::SUCCESS] Deserialization of rdb value failed with error archive_result_t::RANGE_ERROR.
error: Backtrace:
error: Fri Aug 12 13:52:14 2016

       1 [0xb36390]: backtrace_t::backtrace_t() at 0xb36390 (rethinkdb)
       2 [0xb36723]: format_backtrace(bool) at 0xb36723 (rethinkdb)
       3 [0xd9051d]: report_fatal_error(char const*, int, char const*, ...) at 0xd9051d (rethinkdb)
       4 [0x9d6183]: get_data(rdb_value_t const*, buf_parent_t) at 0x9d6183 (rethinkdb)
       5 [0x8cd279]: post_construct_traversal_helper_t::handle_pair(scoped_key_value_t&&, concurrent_traversal_fifo_enforcer_signal_t) at 0x8cd279 (rethinkdb)
       6 [0xab98f1]: concurrent_traversal_adapter_t::handle_pair_coro(scoped_key_value_t*, semaphore_acq_t*, fifo_enforcer_write_token_t, auto_drainer_t::lock_t) at 0xab98f1 (rethinkdb)
       7 [0xab9674]: callable_action_instance_t<std::_Bind<std::_Mem_fn<void (concurrent_traversal_adapter_t::*)(scoped_key_value_t*, semaphore_acq_t*, fifo_enforcer_write_token_t, auto_drainer_t::lock_t)> (concurrent_traversal_adapter_t*, scoped_key_value_t*, semaphore_acq_t*, fifo_enforcer_write_token_t, auto_drainer_t::lock_t)> >::run_action() at 0xab9674 (rethinkdb)
       8 [0xa56fc7]: coro_t::run() at 0xa56fc7 (rethinkdb)
error: Exiting.
```

**Comment by danielmewes** [comment]
Might have found an issue in the backfill code, but have to test more.

**Comment by danielmewes** [comment]
The likely fix for this is now merged into `next` via 78c71d8 and v2.3.x via d1322ad.

It turned out not to be related to backfilling per se, but it was instead an issue in our disk I/o layer.

**Comment by ntquyen** [comment]
Thanks @danielmewes we will upgrade to the new version asap!
